### PR TITLE
adding subnet service endpoints

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Subnet service endpoints are needed for other modules to work, for instance SQL Virtual Network Rule

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
azure_rm_subnet

##### ANSIBLE VERSION
2.7

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py [hunk 1]
```diff
@@ -68,6 +68,19 @@
             - The str can be the name or resource id of the route table.
             - The dict can contains C(name) and C(resource_group) of the route_table.
         version_added: "2.7"
+    service_endpoints:
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_subnet.py:71]: Should we add this as return value in facts module?

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:71]: yes, definitely

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:71]: in fact, it seems like we don't have azure_rm_subnet_facts module....
or i am blind....

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py [hunk 2]
```diff
@@ -68,6 +68,19 @@
...
+        description:
+            - An array of service endpoints.
+        type: list
+        suboptions:
+            service:
+                description:
+                    - The type of the endpoint service.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_subnet.py:78]: required?

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py [hunk 3]
```diff
@@ -257,6 +276,21 @@ def exec_module(self, **kwargs):
                     changed = True
                     results['route_table']['id'] = self.route_table
                     self.log("CHANGED: subnet {0} route_table to {1}".format(self.name, route_table['name']))
+
+                if self.service_endpoints:
+                    oldd = {}
+                    for item in self.service_endpoints:
+                        name = item['service']
+                        oldd[name] = {'service': name, 'locations': item['locations'].sort()}
+                    newd = {}
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_subnet.py:286]: what if service_endpoints not in results?

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:286]: … then change will be detected, as the newd dictionary will be empty

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py (hunk 2 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:78]: yes, added

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py (hunk 3 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:286]: oh, sorry.... needs one fix, thanks for catchin that!

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py [hunk 4]
```diff
@@ -257,6 +277,21 @@ def exec_module(self, **kwargs):
                     changed = True
                     results['route_table']['id'] = self.route_table
                     self.log("CHANGED: subnet {0} route_table to {1}".format(self.name, route_table['name']))
+
+                if self.service_endpoints:
+                    oldd = {}
+                    for item in self.service_endpoints:
+                        name = item['service']
+                        oldd[name] = {'service': name, 'locations': item['locations'].sort()}
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_subnet.py:285]: what's sort order, will it cause any random sort result?

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py [hunk 5]
```diff
@@ -257,6 +277,21 @@ def exec_module(self, **kwargs):
...
+                    newd = {}
+                    if 'service_endpoints' in results:
+                        for item in results['service_endpoints']:
+                            name = item['service']
+                            newd[name] = {'service': name, 'locations': item['locations'].sort()}
+                    if newd != oldd:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: !=  operator works on dict?

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py (hunk 4 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:285]: it's list of strings

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py (hunk 5 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: yes, tested in other modules as well

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: works with nested dicts and lists

Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: no, it's dict newd[name] = {'service': name, 'locations': item['locations'].sort()}

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: yes, dict of services

Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:291]: i don't know. Pls make sure it works.

**Review by yungezz — APPROVED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_subnet.py (hunk 1 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:71]: aha, it is in the virtual network module but in raw format

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_subnet.py:71]: hehe, which we should make obsolete :-)
but anyway, if it's raw, that means this info will be already there, just not documented.
do you think we should have subnet facts?
probably...
I don't think anyone wants to dig up subnet info from virtual network raw format

**Pull request merged**